File: src/netlink/netlink.c
#include <errno.h>
#include <linux/netlink.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#define NETLINK_USER 22
#define USER_MSG 30

#define MAX_PLOAD 1024

int main(int argc, char **argv)
{
    char *data = "hello kernel space by user 21";

    int sockfd, ret;
    struct sockaddr_nl local, remote;

    struct nlmsghdr *nlh = NULL;

    sockfd = socket(AF_NETLINK, SOCK_RAW, USER_MSG);
    if (sockfd == -1)
    {
        printf("create socket failure! %s\n", strerror(errno));
        return -1;
    }

    memset(&local, 0, sizeof(local));
    local.nl_family = AF_NETLINK;
    local.nl_pid = 50;
    local.nl_groups = 0;
    if (bind(sockfd, (struct sockaddr *)&local, sizeof(local)) != 0)
    {
        printf("bind() error!\n");
        close(sockfd);
        return -1;
    }

    memset(&remote, 0, sizeof(remote));
    remote.nl_family = AF_NETLINK;
    remote.nl_pid = 0;
    remote.nl_groups = 0;

    nlh = (struct nlmsghdr *)malloc(NLMSG_SPACE(MAX_PLOAD));
    memset(nlh, 0, sizeof(struct nlmsghdr));
    nlh->nlmsg_len = NLMSG_SPACE(MAX_PLOAD);
    nlh->nlmsg_flags = 0;
    nlh->nlmsg_type = 0;
    nlh->nlmsg_seq = 0;
    nlh->nlmsg_pid = local.nl_pid;

    printf("sendmsg start....\n");

    memcpy(NLMSG_DATA(nlh), data, strlen(data));
    ret = sendto(sockfd, nlh, nlh->nlmsg_len, 0, (struct sockaddr *)&remote,
                 sizeof(struct sockaddr_nl));

    if (ret < 0)
    {
        perror("send to kernel failure!\n");
        close(sockfd);
        exit(-1);
    }

    //接受数据
    printf("recvmsg start....\n");

    memset(nlh, 0, NLMSG_SPACE(MAX_PLOAD));
    ret = recvfrom(sockfd, nlh, NLMSG_LENGTH(MAX_PLOAD), 0, NULL, NULL);
    if (ret < 0)
    {
        printf("recv from kernel failure!\n");
        close(sockfd);
        exit(-1);
    }

    printf("recv data:%s\n", (char *)NLMSG_DATA(nlh));

    close(sockfd);
    free((void *)nlh);

    return 0;
}

File: src/netlink/netlink.go
package main

import (
	"bytes"
	"encoding/binary"
	"fmt"
	"os"
	"syscall"
)

const (
	NETLINK_USER = 22
	USER_MSG     = 29
	MAX_PLOAD    = 1024
)

type MyParam struct {
	ID    int32
	Name  [32]byte
	Value int32
}

func main() {
	// 创建 Netlink 套接字
	sockfd, err := syscall.Socket(syscall.AF_NETLINK, syscall.SOCK_RAW, USER_MSG)
	if err != nil {
		fmt.Printf("create socket failure! %v\n", err)
		os.Exit(1)
	}
	defer syscall.Close(sockfd)

	// 本地地址
	local := &syscall.SockaddrNetlink{
		Family: syscall.AF_NETLINK,
		Pid:    uint32(os.Getpid()),
		Groups: 0,
	}

	// 绑定本地地址
	err = syscall.Bind(sockfd, local)
	if err != nil {
		fmt.Printf("bind() error! %v\n", err)
		os.Exit(1)
	}

	// 远程地址
	remote := &syscall.SockaddrNetlink{
		Family: syscall.AF_NETLINK,
		Pid:    0,
		Groups: 0,
	}

	// 构造消息
	data := MyParam{
		ID:    1,
		Value: 10,
	}
	copy(data.Name[:], "test")

	nlh := &syscall.NlMsghdr{
		Len:   syscall.NLMSG_HDRLEN + uint32(binary.Size(data)),
		Type:  0,
		Flags: 0,
		Seq:   0,
		Pid:   local.Pid,
	}

	// 将消息序列化
	buf := new(bytes.Buffer)
	err = binary.Write(buf, binary.LittleEndian, nlh)
	if err != nil {
		fmt.Printf("binary write error: %v\n", err)
		os.Exit(1)
	}
	err = binary.Write(buf, binary.LittleEndian, data)
	if err != nil {
		fmt.Printf("binary write error: %v\n", err)
		os.Exit(1)
	}

	// 发送消息
	fmt.Println("sendmsg start....")
	err = syscall.Sendto(sockfd, buf.Bytes(), 0, remote)
	if err != nil {
		fmt.Printf("send to kernel failure! %v\n", err)
		os.Exit(1)
	}

	// 接收消息
	fmt.Println("recvmsg start....")
	recvBuf := make([]byte, syscall.Getpagesize())
	n, _, err := syscall.Recvfrom(sockfd, recvBuf, 0)
	if err != nil {
		fmt.Printf("recv from kernel failure! %v\n", err)
		os.Exit(1)
	}

	// 解析接收到的消息
	recvData := recvBuf[syscall.NLMSG_HDRLEN:n]
	fmt.Printf("recv data: %s\n", string(recvData))
}
